### ISSUE TITLE
--terragrunt-include-dir has flaky behavior

### ISSUE THREAD

**Issue by ChristophP** [gruntwork-io/terragrunt#808]
**The bug**
running `terragrunt apply-all` or `destroy-all` with the `--terragrunt-include-dir` flag, can result in different behavior based on the alphabetic ordering of folder names.

**How to reproduce?**
We'll have two test cases with two terragrunt stacks each. In both test folders the first stack has a dependency on the second one, specified in the `terragrunt.hcl`. The only difference between the test folders is that `test1` has a folder called `second` and `test2` has a folder called `_second`. The underscore makes it be alphabetically sorted before `first` whereas without it, it would come in, well, second.
```.
.
|----module
| |----main.tf
|----test1
| |----second
| | |----terragrunt.hcl
| |----first
| | |----terragrunt.hcl
|----test2
| |----_second     <- note the leading underscore here
| | |----terragrunt.hcl
| |----first
| | |----terragrunt.hcl
```
I prepared such a test repo [here](https://github.com/ChristophP/terragrunt-include-bug-example).
We run terragrunt in both directories from the root of the directory structure and get different results:
```sh
# test 1
terragrunt apply-all --terragrunt-working-dir test1 --terragrunt-include-dir first
  => Module /home/terragrunt-test/test1/first (excluded: false, dependencies: [/home/terragrunt-test/test1/second])
  => Module /home/terragrunt-test/test1/second (excluded: true, dependencies: [])
```
```sh
# test  2
terragrunt apply-all --terragrunt-working-dir test2 --terragrunt-include-dir first
  => Module /home/terragrunt-test/test2/_second (excluded: false, dependencies: [])
  => Module /home/terragrunt-test/test2/first (excluded: false, dependencies: [/home/terragrunt-test/test2/_second])
```

**Comment by ChristophP** [comment]
This could be fixed by making a changing that part in `configstack/module.go` [here](https://github.com/gruntwork-io/terragrunt/blob/master/configstack/module.go#L182-L193)

I am just not sure what the desired behavior is here. Include dependencies of included directories? That seems like a bad default especially for `destroy`.

Can anyone share their thoughts on this?

**Comment by brikis98** [comment]
I'm not sure I fully understand the bug. Why does the leading underscore / alphabetical order matter here?

As a general note, the way Terragrunt handles dependencies needs a lot of work. See https://github.com/gruntwork-io/terragrunt/issues/765, https://github.com/gruntwork-io/terragrunt/issues/272, and https://github.com/gruntwork-io/terragrunt/issues/524. Ideas and PRs to improve that (and fix this bug) are very welcome!

**Comment by ChristophP** [comment]
Hi @brikis98,

I am going to assume the following in my response:
*The desired behavior currently is to include all folders mentioned in the include flag AS WELL as their dependencies*

> I'm not sure I fully understand the bug. Why does the leading underscore / alphabetical order matter here?

**tl;dr**
The current code behaves in a way that it excludes previously included modules, which is undesirable. It shouldn't matter in which order the modules are processed, the exclusion/inclusion result should be the same. But that's not what the current implementation does. As you see above in the example, if we do a folder rename `second -> _second` the folder gets falsely excluded even though it is a dependency of the folder `first`.

**detailed explanation**
In pseudo code.
Case 1 (folders: `first` and `second`. where `first` has `second` as a dependency)
```
modules are [first, second]
first iteration:
  is "first" mentioned in include flags?
    YES -> include it (exclude = false)
           proceed to include it's dependencies, that means: include "second"
second iteration:
  is "second" mentioned in include flags?
    NO -> exclude it (exclude = true)
          "second" has no dependencies, so nothing left to do

RESULT: first (exluded =  false), second (excluded = true)
```
Case 2 (folders: `first` and `_second`. where `first` has `_second` as a dependency)
```
modules are [_second, first]
first iteration:
  is "_second" mentioned in include flags?
    NO -> exclude it (exclude = true)
          "_second" has no dependencies, so nothing left to do
second iteration:
  is "first" mentioned in include flags?
    YES -> include it (exclude = false)
           proceed to include it's dependencies, that means: include "_second"

RESULT: first (exluded =  false), second (excluded = false)
```

Does that make it clear?

**Comment by ChristophP** [comment]
@brikis98 I didn't hear back on this issue and I am not sure where this ranks on the priority list but I wanted to make the bug clearer once more.
When running `terragrunt apply-all --terragrunt-include-dir firstDir` where `firstDir` has a dependency `secondDir` both directories will be included. However, when `secondDir` is renamed to `_secondDir` it will NOT be included (as so now `firstDir` depends on `_secondDir`) because of the way the implementation works at the moment. Same thing different outcome. Only difference being a name of a directory.

**Comment by brikis98** [comment]
Thanks for the detailed explanation! Does indeed seem like a bug. A PR to fix it is very welcome!
